src/reader/reader_handler.py

```diff
@@ -14,53 +17,101 @@ def __init__(self):
-        for entry in filenames:
-            # Create full path
-            fullPath = os.path.join(self.data_dir, entry)
-            files.append(fullPath)
+        fullpaths = [os.path.join(self.data_dir, entry) for entry in filenames]
+
+        # Load the list of files to read
+        # We use a held-out data set for evaluation
+        # Note: this could also be done in __init__() so as not to repeat the work
+        if self.eval_enabled and do_eval:
+            files = [path for i, path in enumerate(fullpaths) if i in self.eval_indices]
+            num_files = self.num_files_eval
+            self.batch_size = self.batch_size_eval
+        else:
+            if self.eval_enabled:
+                files = [path for i, path in enumerate(fullpaths) if i not in self.eval_indices]
+            else:
+                files = fullpaths
+            num_files = self.num_files_train
+            self.batch_size = self.batch_size_train
+
+        assert len(files) == num_files, f"Expected {num_files} training files but {len(files)} found. Ensure data was generated correctly."
+
+        read_shuffle = True
+        if self.read_shuffle == Shuffle.OFF:
+            read_shuffle = False
+
         seed = None
-        if FileAccess.MULTI == self.file_access:
-            files = files[:self.num_files]
-            read_shuffle = True
-            if self.read_shuffle == Shuffle.OFF:
-                read_shuffle = False
-            if read_shuffle:
-                seed = self.seed
-                if self.seed_change_epoch:
-                    seed = self.seed + epoch_number
+        if read_shuffle:
+            seed = self.seed
+            if self.seed_change_epoch:
+                seed = self.seed + epoch_number
+
+        if seed is not None:
+            random.seed(seed)
+
+        if self.framework.get_type() == FrameworkType.TENSORFLOW and FileAccess.MULTI == self.file_access:
+            # Partition the files for each rank
             partition_size = int(math.ceil(len(files) / self.comm_size))
             part_start, part_end = (partition_size * self.my_rank, partition_size * ( self.my_rank + 1))
             self._local_file_list = files[part_start:part_end]
-            print("rank {}, file_list {}, size {}".format(self.my_rank, self._local_file_list,partition_size))
-            if seed is not None:
-                random.seed(seed)
+            self._local_file_list_size = len(self._local_file_list)
+
             if read_shuffle:
                 random.shuffle(self._local_file_list)
+
+            logging.debug(f"{utcnow()} Rank {self.my_rank} will read {self._local_file_list_size} files: {self._local_file_list}")
+
+        # For PyTorch the file partitioning is done by the DistributedSampler in src/reader/data_loader_reader.py.
+        # We could also partition the files here like for TF and not use the sampler.
+        # We did it this way just to follow the UNET3D implementation as closely as possible.
```

Comment:
I think this is interesting. We could decide how and when we want to partition based on a flag potentially to keep both versions and support more cases.

Reply:
Maybe add a issue for this as enhancement.